
# DESeq Differential Expression Analysis Results

> M. I. Love, W. Huber, S. Anders: Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology 2014, 15:550.

## DESeq Result Table

* `r r_data$deseq_results@elementMetadata$description[4]`

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
if(is.null(r_data$deseq_results)) return()
    tbl<-as.data.frame(r_data$deseq_results)
    if(nrow(tbl) == 0) return()
    DT::datatable(tbl, selection = 'single', options = list(scrollX = TRUE, scrollY = "250px", searching=T, order = list(list(6, 'asc')) , orderClasses = T))
```

### Summary

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
    if(is.null(r_data$deseq_results)) return()
    sm <- capture.output(DESeq2::summary.DESeqResults(r_data$deseq_results))
    
```

* `r paste0("Total number of significant genes: ", sum(r_data$deseq_results$padj < input$deseq_alpha, na.rm = T), ".")`
    
* `r sm[4]`
    
* `r sm[5]`
    
* `r sm[6]`
    
* `r paste(sm[7], sm[8])`


## DESeq MA Plot

```{r echo = FALSE, warning=FALSE, message = FALSE, fig.width = 10, fig.height = 8}
if(is.null(r_data$deseq_results)) return()
BiocGenerics::plotMA(r_data$deseq_results, main="DESeq2", ylim=c(-2,2))
```


## DESeq Gene Expression Plot

```{r echo = FALSE, warning=FALSE, message = FALSE}
    if(is.null(r_data$deseq_results)) return()
    s = input$deseq_result_tbl_row_last_clicked
    tbl<-as.data.frame(r_data$deseq_results)

    if (length(s)) {
        selected_gene <- rownames(tbl[s, , drop = FALSE])
        d <- as.data.frame(t(r_data$df[selected_gene,])) %>% tibble::rownames_to_column()
        colnames(d) <- c("sample", "expression_level")

        if(r_data$deseq_params$test == "LRT") {
            samples = NULL
        } else {
            req(input$deseq_group1, input$deseq_group2)
            samples = r_data$meta[,1][which(r_data$meta[,r_data$deseq_params$condition] %in% c(input$deseq_group1, input$deseq_group2))]
        }

        callModule(pivot_featurePlot, "deseq_gene_plt", meta = r_data$meta, df = d, gene = selected_gene, ids = samples)
    } else {
        return()
    }
    
```


